controllers/batchrelease/workloads/cloneset_control_plan.go

```diff
@@ -0,0 +1,334 @@
+	}
+
+	// record revisions and replicas info to BatchRelease.Status
+	c.recordCloneSetRevisionAndReplicas()
+
+	c.recorder.Event(c.parentController, v1.EventTypeNormal, "InitializedSuccessfully", "Rollout resource are initialized")
+	return true, nil
+}
+
+// ProgressOneBatchReplicas calculates the number of pods we can upgrade once according to the rollout spec
+// and then set the partition accordingly
+func (c *CloneSetRolloutController) ProgressOneBatchReplicas() (bool, error) {
+	if err := c.fetchCloneSet(); err != nil {
+		return false, nil
+	}
+
+	// if the workload status is untrustworthy
+	if c.clone.Status.ObservedGeneration != c.clone.Generation {
+		return false, nil
+	}
+
+	// the number of canary pods should have in current batch
+	canaryGoal := c.calculateCurrentCanary(c.releaseStatus.ObservedWorkloadReplicas)
+	// the number of stable pods should have in current batch
+	stableGoal := c.calculateCurrentStable(c.releaseStatus.ObservedWorkloadReplicas)
+	// the number of canary pods now we have in current state
+	currentCanaryReplicas := c.clone.Status.UpdatedReplicas
+	// workload partition calculated
+	workloadPartition, _ := intstr.GetValueFromIntOrPercent(c.clone.Spec.UpdateStrategy.Partition,
+		int(c.releaseStatus.ObservedWorkloadReplicas), true)
+
+	// in case of no need to upgrade pods
+	if currentCanaryReplicas >= canaryGoal && int32(workloadPartition) <= stableGoal {
+		klog.V(3).InfoS("upgraded one batch, but no need to update partition of cloneset",
+			"BatchRelease", c.releasePlanKey, "current-batch", c.releaseStatus.CanaryStatus.CurrentBatch,
+			"canary-goal", canaryGoal, "stable-goal", stableGoal, "canary-replicas", currentCanaryReplicas, "partition", workloadPartition)
+		return true, nil
+	}
+
+	// upgrade pods
+	if err := c.patchCloneSetPartition(c.clone, stableGoal); err != nil {
+		return false, nil
+	}
+
+	klog.V(3).InfoS("upgraded one batch", "BatchRelease", c.releasePlanKey,
+		"current batch", c.releaseStatus.CanaryStatus.CurrentBatch, "updateRevision size", canaryGoal)
+	c.recorder.Eventf(c.parentController, v1.EventTypeNormal, "SetBatchDone",
+		"Finished submitting all upgrade quests for batch %d", c.releaseStatus.CanaryStatus.CurrentBatch)
+	return true, nil
+}
+
+// CheckOneBatchReplicas checks to see if the pods are all available according to the rollout plan
+func (c *CloneSetRolloutController) CheckOneBatchReplicas() (bool, error) {
+	if err := c.fetchCloneSet(); err != nil {
+		return false, nil
+	}
+
+	// if the workload status is untrustworthy
+	if c.clone.Status.ObservedGeneration != c.clone.Generation {
+		return false, nil
+	}
+
+	// the number of canary pods now we have in current state
+	canaryReplicas := c.clone.Status.UpdatedReplicas
+	// the number of stable pods now we have in current state
+	stableReplicas := c.clone.Status.Replicas - canaryReplicas
+	// the number of canary pods that have been ready in current state
+	canaryReadyReplicas := c.clone.Status.UpdatedReplicas
```

Comment:
c.clone.Status.UpdatedReadyReplicas is the updated ready replicas

Reply:
done